File: backend-springboot/src/main/java/com/monitor/controller/BehaviorController.java
package com.monitor.controller;

import com.monitor.service.BehaviorService;
import com.monitor.service.ElasticsearchService;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.HashMap;
import java.util.Map;

/**
 * 用户行为控制器
 */
@RestController
@RequestMapping("/behavior")
@RequiredArgsConstructor
public class BehaviorController {
    
    private final BehaviorService behaviorService;
    private final ElasticsearchService elasticsearchService;
    
    @Value("${default-project-id:001}")
    private String defaultProjectId;
    
    @PostMapping("/report")
    public ResponseEntity<Map<String, Object>> report(@RequestBody Map<String, Object> body) {
        return handleReport(body);
    }
    
    @GetMapping("/report")
    public ResponseEntity<Map<String, Object>> reportGet(@RequestParam String data) {
        try {
            String decoded = java.net.URLDecoder.decode(data, java.nio.charset.StandardCharsets.UTF_8);
            com.fasterxml.jackson.databind.ObjectMapper mapper = new com.fasterxml.jackson.databind.ObjectMapper();
            Map<String, Object> body = mapper.readValue(decoded, Map.class);
            return handleReport(body);
        } catch (Exception e) {
            System.err.println("Behavior report GET failed: " + e.getMessage());
            return ResponseEntity.ok(Map.of("success", false, "message", "数据解析失败"));
        }
    }
    
    private ResponseEntity<Map<String, Object>> handleReport(Map<String, Object> body) {
        try {
            // 调试日志：记录路由变化和自定义事件的 userId
            String type = (String) body.getOrDefault("type", "pv");
            if ("route-change".equals(type) || "custom".equals(type)) {
                System.out.println("📊 " + type + " event - userId: " + body.get("userId") + ", sessionId: " + body.get("sessionId"));
            }
            
            @SuppressWarnings("unchecked")
            Map<String, Object> data = (Map<String, Object>) body.get("data");
            
            behaviorService.report(
                defaultProjectId,
                type,
                (String) body.getOrDefault("url", ""),
                (String) body.get("path"),
                (String) body.get("userId"),
                (String) body.get("sessionId")
            );
            
            // 异步写入 Elasticsearch（不阻塞主流程）
            Map<String, Object> logData = new HashMap<>();
            logData.put("projectId", defaultProjectId);
            logData.put("type", "behavior");
            logData.put("userId", body.get("userId"));
            logData.put("sessionId", body.get("sessionId"));
            logData.put("url", body.getOrDefault("url", ""));
            logData.put("path", body.get("path"));
            logData.put("timestamp", body.getOrDefault("timestamp", new java.util.Date().toInstant().toString()));
            logData.put("behaviorType", type);
            
            String message;
            if ("route-change".equals(type)) {
                message = "Route change: " + (data != null ? data.get("from") : "") + " -> " + (data != null ? data.get("to") : "");
            } else if ("custom".equals(type)) {
                message = "Custom event: " + (data != null ? data.get("eventName") : "");
            } else {
                message = type;
            }
            logData.put("message", message);
            
            if (data != null) {
                logData.putAll(data);
            }
            
            try {
                elasticsearchService.writeLog(logData);
            } catch (Exception e) {
                System.out.println("⚠️ Elasticsearch write failed (non-blocking): " + e.getMessage());
            }
            
            return ResponseEntity.ok(Map.of("success", true));
        } catch (Exception e) {
            System.err.println("Behavior report failed: " + e.getMessage());
            return ResponseEntity.ok(Map.of("success", false, "message", "上报失败"));
        }
    }
    
    @GetMapping("/stats")
    public ResponseEntity<Map<String, Object>> stats(@RequestParam(required = false) String type) {
        Map<String, Object> stats = behaviorService.getStats(defaultProjectId, type);
        return ResponseEntity.ok(Map.of("success", true, "data", stats));
    }
    
    @GetMapping("/events")
    public ResponseEntity<Map<String, Object>> getEvents(
            @RequestParam(required = false) String type,
            @RequestParam(required = false, defaultValue = "1") Integer page,
            @RequestParam(required = false, defaultValue = "20") Integer pageSize) {
        // TODO: 实现行为事件详情列表查询
        return ResponseEntity.ok(Map.of(
            "success", true,
            "data", Map.of(
                "list", java.util.List.of(),
                "total", 0,
                "page", page,
                "pageSize", pageSize
            )
        ));
    }
}

File: backend-midway/src/controller/behavior.controller.ts
import { Controller, Get, Post, Inject, Body, Query, Config } from '@midwayjs/core';
import { ApiTags, ApiOperation, ApiQuery, ApiBody } from '@midwayjs/swagger';
import { BehaviorService } from '../service/behavior.service';
import { ElasticsearchService } from '../service/elasticsearch.service';

@Controller('/behavior')
@ApiTags(['behavior'])
export class BehaviorController {
  @Inject()
  behaviorService: BehaviorService;

  @Inject()
  elasticsearchService: ElasticsearchService;

  @Config('defaultProjectId')
  defaultProjectId: string;

  @Post('/report')
  @ApiOperation({ summary: '上报用户行为数据' })
  @ApiBody({ description: '用户行为数据' })
  async report(@Body() body: any) {
    return this.handleReport(body);
  }

  @Get('/report')
  @ApiOperation({ summary: '上报用户行为数据（GET方式，用于img上报）' })
  @ApiQuery({ name: 'data', required: true, description: 'JSON编码的行为数据' })
  async reportGet(@Query('data') dataStr: string) {
    try {
      const body = JSON.parse(decodeURIComponent(dataStr));
      return this.handleReport(body);
    } catch (error) {
      console.error('Behavior report GET failed:', error);
      return { success: false, message: '数据解析失败' };
    }
  }

  private async handleReport(body: any) {
    try {
      // 调试日志：记录路由变化和自定义事件的 userId
      if (body.type === 'route-change' || body.type === 'custom') {
        console.log(`📊 ${body.type} event - userId: ${body.userId || 'undefined'}, sessionId: ${body.sessionId || 'undefined'}`);
      }
      
      await this.behaviorService.report({
        projectId: this.defaultProjectId,
        type: body.type || 'pv',
        url: body.url || '',
        path: body.path,
        userId: body.userId,
        sessionId: body.sessionId,
        data: body.data, // 传递详细数据
      });

      // 异步写入 Elasticsearch（不阻塞主流程）
      this.elasticsearchService.writeLog({
        projectId: this.defaultProjectId,
        type: 'behavior',
        userId: body.userId,
        sessionId: body.sessionId,
        url: body.url || '',
        path: body.path,
        timestamp: body.timestamp || new Date().toISOString(),
        behaviorType: body.type || 'pv',
        message: body.type === 'route-change' 
          ? `Route change: ${body.data?.from || ''} -> ${body.data?.to || ''}`
          : body.type === 'custom'
          ? `Custom event: ${body.data?.eventName || ''}`
          : body.type || 'pv',
        ...body.data,
      }).catch((err) => {
        // Elasticsearch 写入失败不影响主流程，只记录错误
        console.warn('⚠️ Elasticsearch write failed (non-blocking):', err.message);
      });

      return { success: true };
    } catch (error) {
      console.error('Behavior report failed:', error);
      return { success: false, message: '上报失败' };
    }
  }

  @Get('/stats')
  @ApiOperation({ summary: '获取用户行为统计' })
  @ApiQuery({ name: 'type', required: false, description: '行为类型：pv, click, route-change, custom' })
  async stats(@Query('type') type?: string) {
    const stats = await this.behaviorService.getStats(this.defaultProjectId, type);
    return { success: true, data: stats };
  }

  @Get('/events')
  @ApiOperation({ summary: '获取行为事件详情列表（点击事件和自定义事件）' })
  @ApiQuery({ name: 'type', required: false, description: '行为类型：click, custom' })
  @ApiQuery({ name: 'page', required: false, description: '页码，从1开始' })
  @ApiQuery({ name: 'pageSize', required: false, description: '每页数量' })
  async getEvents(
    @Query('type') type?: string,
    @Query('page') page?: number,
    @Query('pageSize') pageSize?: number
  ) {
    const events = await this.behaviorService.getEvents(
      this.defaultProjectId,
      type,
      page || 1,
      pageSize || 20
    );
    return { success: true, data: events };
  }
}

